### PR TITLE
[FE] charts & refactoring & some devops

### PR DESCRIPTION
**Pull request by mwkim0919**
* charts using real data
* some refactoring done (mostly in `budgetUtil.ts`)
* devops
  * running prettier and tests before commit (FE)

### CODE REVIEW

File: frontend/src/utils/test/budgetUtil.test.ts [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    expect(depositWithdrawMaps['2018-08']).toHaveProperty(['deposit'], 0);
+    expect(depositWithdrawMaps['2018-07']).toHaveProperty(['deposit'], 0);
+    expect(depositWithdrawMaps['2018-06']).toHaveProperty(['deposit'], 0);
+    expect(depositWithdrawMaps['2018-05']).toHaveProperty(['deposit'], 0);
+    expect(depositWithdrawMaps['2018-04']).toHaveProperty(['deposit'], 0);
+    expect(depositWithdrawMaps['2018-03']).toHaveProperty(['deposit'], 0);
+    expect(depositWithdrawMaps['2018-02']).toHaveProperty(['deposit'], 8642.24);
+    expect(depositWithdrawMaps['2018-01']).toHaveProperty(['deposit'], 8642.24);
+
+    // expect(depositWithdrawMaps['2018-11']).toHaveProperty(['withdraw'], 325.35);
```
Comment (mwkim0919) [inline frontend/src/utils/test/budgetUtil.test.ts]: floating point issue. Need to figure out how to check this.

File: frontend/src/utils/test/budgetUtil.test.ts [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    expect(depositWithdrawMaps['2018-10']).toHaveProperty(['withdraw'], 14.66);
+    expect(depositWithdrawMaps['2018-09']).toHaveProperty(['withdraw'], 148.50);
+    expect(depositWithdrawMaps['2018-08']).toHaveProperty(['withdraw'], 148.50);
+    expect(depositWithdrawMaps['2018-07']).toHaveProperty(['withdraw'], 148.50);
+    expect(depositWithdrawMaps['2018-06']).toHaveProperty(['withdraw'], 148.50);
+    expect(depositWithdrawMaps['2018-05']).toHaveProperty(['withdraw'], 148.50);
+    expect(depositWithdrawMaps['2018-04']).toHaveProperty(['withdraw'], 98);
+    // expect(depositWithdrawMaps['2018-03']).toHaveProperty(['withdraw'], 34.38);
```
Comment (mwkim0919) [inline frontend/src/utils/test/budgetUtil.test.ts]: floating point issue. Need to figure out how to check this.

File: frontend/src/components/Budget/Transaction/TransactionHome.tsx [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+import { Transaction, Category } from 'src/types/budget';
+
+interface Props {
+  transactions: Transaction[];
+  categories: Category[];
+}
+
+class TransactionHome extends React.Component<Props> {
+  render() {
+    // const { transactions, categories } = this.props;
```
Comment (mwkim0919) [inline frontend/src/components/Budget/Transaction/TransactionHome.tsx]: Couldn't figure out how to connect this.

File: frontend/src/utils/budgetUtil.ts [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    const keyArray = [selectedCategory.name];
+    while (selectedCategory.parentCategory) {
+      keyArray.push(selectedCategory.parentCategory.name);
+      selectedCategory = selectedCategory.parentCategory;
+    }
+    return keyArray.reverse();
+  }
+}
+
+function getParentCategoryNameById(id: number, categories: Category[]): string {
```
Comment (johnp418) [inline frontend/src/utils/budgetUtil.ts]: Same logic as previous comments. Take category as argument and you don't need to iterate.

File: frontend/package.json [hunk 5]
```diff
@@ -75,9 +77,15 @@
   },
   "husky": {
     "hooks": {
-      "pre-push": "yarn run test"
+      "pre-commit": "lint-staged && yarn run test"
     }
   },
+  "lint-staged": {
```
Comment (johnp418) [inline frontend/package.json]: 👍

File: frontend/src/utils/budgetUtil.ts [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+  [categoryName: string]: number;
+}
+
+export const generateCategoryMaps = (
+  categories: Category[],
+  transactions: Transaction[]
+) => {
+  const result = initializeCategoryMaps(categories);
+  for (const transaction of transactions) {
+    const categoryNames = getCategoryNames(transaction.categoryId, categories);
```
Comment (johnp418) [inline frontend/src/utils/budgetUtil.ts]: I'd create a map of categories before entering the loop and pass the category here directly. This way, you don't have to iterate inside the callee.
```suggestion
    const categoryNames = getCategoryNames(categoriesMap[transaction.categoryId]);
```

File: frontend/src/utils/budgetUtil.ts [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+export const generateParentCategoryWithdrawMaps = (
+  transactions: Transaction[],
+  categories: Category[]
+): CategoryWithdrawMap => {
+  return transactions.reduce(
+    (acc: CategoryWithdrawMap, transaction: Transaction) => {
+      const amount = transaction.withdraw ? Number(transaction.withdraw) : 0;
+      const categoryName = transaction.categoryId
+        ? getParentCategoryNameById(transaction.categoryId, categories)
```
Comment (johnp418) [inline frontend/src/utils/budgetUtil.ts]: Same as above.
Pass category as argument and you don't need to iterate in `getParentCategoryNameById` function.

File: frontend/src/utils/budgetUtil.ts [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+export interface MonthlyDepositWithdraw {
+  [month: string]: DepositWithdraw;
+}
+
+export interface DepositWithdraw {
+  deposit: number;
+  withdraw: number;
+}
+
+export interface CategoryMap {
```
Comment (johnp418) [inline frontend/src/utils/budgetUtil.ts]: It makes me think this type is a map with key as categoryId and category as value. Let's avoid this type of name unless it's really a map of category.

File: frontend/src/utils/test/budgetUtil.test.ts (hunk 1 above)
Comment (johnp418) [reply frontend/src/utils/test/budgetUtil.test.ts]: Let's discuss this week.

**Comment by johnp418** [comment]
It's so easier to read than before. You needs double thumbs up

File: frontend/src/utils/test/budgetUtil.test.ts (hunk 2 above)
Comment (mwkim0919) [reply frontend/src/utils/test/budgetUtil.test.ts]: Fixed but probably not the best way.

File: frontend/src/components/Budget/Transaction/TransactionHome.tsx (hunk 3 above)
Comment (mwkim0919) [reply frontend/src/components/Budget/Transaction/TransactionHome.tsx]: Did this taking another approach: connecting `TransactionHomeContainer`

File: frontend/package.json (hunk 5 above)
Comment (mwkim0919) [reply frontend/package.json]: :+1:

File: frontend/src/utils/budgetUtil.ts [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+function initializeCategoryMaps(categories: Category[]): CategoryMaps {
+  const result = {
+    Others: {
+      expense: 0,
+      limit: 0,
+    },
+  };
+  for (const category of categories) {
+    let categoryTemp = category;
+    const keyArray = [categoryTemp.name];
```
[the remark this answers is not in the document]
Comment (mwkim0919) [reply frontend/src/utils/budgetUtil.ts]: Refactored. Thanks!

File: frontend/src/utils/budgetUtil.ts [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+): string[] {
+  if (categoryId === null) {
+    return ['Others'];
+  }
+  let selectedCategory = categories.filter(
+    category => category.id === categoryId
+  )[0];
+  return putRelatedCategoryNamesInArray();
+
+  function putRelatedCategoryNamesInArray(): string[] {
```
[the remark this answers is not in the document]
Comment (mwkim0919) [reply frontend/src/utils/budgetUtil.ts]: Got it.

File: frontend/src/utils/budgetUtil.ts (hunk 4 above)
Comment (mwkim0919) [reply frontend/src/utils/budgetUtil.ts]: Changed.

File: frontend/src/utils/budgetUtil.ts (hunk 6 above)
Comment (mwkim0919) [reply frontend/src/utils/budgetUtil.ts]: Changed.

File: frontend/src/utils/budgetUtil.ts (hunk 8 above)
Comment (mwkim0919) [reply frontend/src/utils/budgetUtil.ts]: Changed.

File: frontend/src/utils/budgetUtil.ts (hunk 6 above)
Comment (mwkim0919) [reply frontend/src/utils/budgetUtil.ts]: Changed

**Pull request merged**